File: apps/macos/CodexBridgeApp/Domain/CaptureValidator.swift
import CryptoKit
import Foundation

enum CaptureValidationError: LocalizedError, Equatable {
    case unsupportedVersion(Int)
    case unsupportedSource(String)
    case invalidHost
    case emptyConversation
    case oversized(Int)
    case turnOrder
    case duplicateMessageID(String)
    case invalidSelection
    case inconsistentTurn(String)

    var errorDescription: String? {
        switch self {
        case .unsupportedVersion: "这个对话文件由其他版本生成，请更新 Codex Bridge 后重试。"
        case .unsupportedSource, .invalidHost: "只支持导入由 Codex Bridge 浏览器扩展保存的 ChatGPT 对话。"
        case .emptyConversation: "这个文件中没有可导入的对话内容。"
        case .oversized: "对话内容超过 2 MB。请减少选择的内容后重试。"
        case .turnOrder, .duplicateMessageID, .invalidSelection, .inconsistentTurn:
            "对话文件内容不完整，请回到 ChatGPT 网页重新选择并保存。"
        }
    }
}

struct CaptureValidator: Sendable {
    static let maximumPayloadBytes = 2 * 1_024 * 1_024

    func validate(data: Data) throws -> CapturedConversation {
        guard data.count <= Self.maximumPayloadBytes else {
            throw CaptureValidationError.oversized(data.count)
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let payload = try decoder.decode(CapturePayload.self, from: data)
        return try validate(payload)
    }

    func validate(_ payload: CapturePayload) throws -> CapturedConversation {
        guard payload.schemaVersion == 1 else {
            throw CaptureValidationError.unsupportedVersion(payload.schemaVersion)
        }
        guard payload.source.kind == ConversationSourceKind.chatGPTWeb.rawValue else {
            throw CaptureValidationError.unsupportedSource(payload.source.kind)
        }
        guard payload.source.url.host?.lowercased() == "chatgpt.com" else {
            throw CaptureValidationError.invalidHost
        }
        guard !payload.turns.isEmpty else {
            throw CaptureValidationError.emptyConversation
        }
        guard payload.turns.map(\.index) == payload.turns.map(\.index).sorted(),
              Set(payload.turns.map(\.index)).count == payload.turns.count else {
            throw CaptureValidationError.turnOrder
        }

        var messageIDs = Set<String>()
        for turn in payload.turns {
            for message in [turn.user, turn.assistant].compactMap({ $0 }) {
                guard messageIDs.insert(message.id).inserted else {
                    throw CaptureValidationError.duplicateMessageID(message.id)
                }
            }
            guard !turn.user.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                throw CaptureValidationError.inconsistentTurn(turn.id)
            }
            if turn.complete, turn.assistant == nil {
                throw CaptureValidationError.inconsistentTurn(turn.id)
            }
        }

        let availableIDs = payload.turns.map(\.id)
        guard payload.selection.selectedTurnIds == availableIDs else {
            throw CaptureValidationError.invalidSelection
        }

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        let canonicalData = try encoder.encode(payload.turns)
        let hash = SHA256.hash(data: canonicalData).map { String(format: "%02x", $0) }.joined()

        var warnings = payload.warnings
        if payload.capture.attachmentCount > 0 {
            warnings.append("检测到 \(payload.capture.attachmentCount) 个附件。附件不会随对话一起保存，请在需要时从原会话添加。")
        }
        if !payload.capture.complete {
            warnings.append("部分较早的消息可能未包含。请在 ChatGPT 中向上滚动后重新保存。")
        }

        let sourceConversationID = payload.source.conversationId?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .nilIfEmpty
        let identitySeed = sourceConversationID.map { "chatgpt-web-\($0)" }
            ?? "chatgpt-web-content-\(hash)"

        return CapturedConversation(
            id: CapturedConversation.stableID(for: identitySeed),
            sourceKind: .chatGPTWeb,
            sourceURL: payload.source.url,
            sourceConversationID: sourceConversationID,
            title: payload.source.title,
            projectName: nil,
            projectPath: nil,
            capturedAt: payload.capture.capturedAt,
            captureScope: payload.capture.scope,
            freshness: .captured,
            contentHash: hash,
            turns: payload.turns,
            warnings: warnings,
            codexThreadID: nil
        )
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: apps/windows/CodexBridge.Core/Capture/CaptureValidator.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;

namespace CodexBridge.Core.Capture;

public sealed class CaptureValidator
{
    public const int MaximumPayloadBytes = 2 * 1_024 * 1_024;

    public CaptureValidationResult Validate(ReadOnlySpan<byte> utf8Json)
    {
        if (utf8Json.Length > MaximumPayloadBytes)
        {
            throw Error(
                CaptureValidationErrorCode.Oversized,
                $"Capture payload exceeds the {MaximumPayloadBytes}-byte limit.");
        }

        CapturePayload payload;
        try
        {
            payload = CaptureJson.Deserialize(utf8Json);
        }
        catch (Exception exception) when (exception is JsonException or NotSupportedException)
        {
            throw new CaptureValidationException(
                CaptureValidationErrorCode.InvalidPayload,
                "Capture payload is not valid JSON.",
                exception);
        }

        return Validate(payload);
    }

    public CaptureValidationResult Validate(CapturePayload payload)
    {
        ArgumentNullException.ThrowIfNull(payload);

        if (payload.SchemaVersion != 1)
        {
            throw Error(CaptureValidationErrorCode.UnsupportedVersion, "Unsupported capture schema version.");
        }

        if (payload.Source is null || payload.Source.Kind != "chatgpt-web")
        {
            throw Error(CaptureValidationErrorCode.UnsupportedSource, "Unsupported capture source.");
        }

        if (payload.Source.Url is null
            || !payload.Source.Url.IsAbsoluteUri
            || !string.Equals(payload.Source.Url.Host, "chatgpt.com", StringComparison.OrdinalIgnoreCase))
        {
            throw Error(CaptureValidationErrorCode.InvalidHost, "Capture URL must belong to chatgpt.com.");
        }

        if (string.IsNullOrEmpty(payload.Source.Title) || payload.Source.Title.Length > 500)
        {
            throw Error(CaptureValidationErrorCode.InvalidPayload, "Capture source title is invalid.");
        }

        if (payload.Capture is null
            || payload.Capture.Scope != "rendered-current-page"
            || payload.Capture.AttachmentCount < 0)
        {
            throw Error(CaptureValidationErrorCode.InvalidPayload, "Capture metadata is invalid.");
        }

        if (payload.Warnings is null || payload.Warnings.Any(warning => warning is null))
        {
            throw Error(CaptureValidationErrorCode.InvalidPayload, "Capture warnings are invalid.");
        }

        if (payload.Turns is null || payload.Turns.Count == 0)
        {
            throw Error(CaptureValidationErrorCode.EmptyConversation, "Capture contains no conversation turns.");
        }

        ValidateTurnOrder(payload.Turns);
        ValidateMessages(payload.Turns);
        ValidateSelection(payload.Selection, payload.Turns);

        var canonicalTurns = SerializeCanonicalTurns(payload.Turns);
        var contentHash = Convert.ToHexString(SHA256.HashData(canonicalTurns)).ToLowerInvariant();
        var conversationId = payload.Source.ConversationId?.Trim();
        var identitySeed = string.IsNullOrEmpty(conversationId)
            ? $"chatgpt-web-content-{contentHash}"
            : $"chatgpt-web-{conversationId}";

        return new CaptureValidationResult(
            payload,
            CreateStableGuid(identitySeed),
            contentHash);
    }

    private static void ValidateTurnOrder(IReadOnlyList<CapturedTurn> turns)
    {
        var indexes = new HashSet<int>();
        var previousIndex = int.MinValue;
        foreach (var turn in turns)
        {
            if (turn is null
                || string.IsNullOrEmpty(turn.Id)
                || turn.Index < 0
                || turn.Index < previousIndex
                || !indexes.Add(turn.Index))
            {
                throw Error(CaptureValidationErrorCode.TurnOrder, "Capture turn order is invalid.");
            }

            previousIndex = turn.Index;
        }
    }

    private static void ValidateMessages(IReadOnlyList<CapturedTurn> turns)
    {
        var messageIds = new HashSet<string>(StringComparer.Ordinal);
        foreach (var turn in turns)
        {
            if (turn.User is null
                || string.IsNullOrEmpty(turn.User.Id)
                || turn.User.IdSource is null
                || turn.User.IdSource is not ("dom" or "dom-order")
                || string.IsNullOrEmpty(turn.User.Text))
            {
                throw Error(CaptureValidationErrorCode.InvalidPayload, "Capture contains an invalid user message.");
            }

            if (!messageIds.Add(turn.User.Id))
            {
                throw Error(
                    CaptureValidationErrorCode.DuplicateMessageId,
                    $"Duplicate message ID: {turn.User.Id}");
            }

            if (string.IsNullOrWhiteSpace(turn.User.Text))
            {
                throw Error(CaptureValidationErrorCode.InconsistentTurn, $"Turn {turn.Id} has no user text.");
            }

            if (turn.Assistant is not null)
            {
                if (string.IsNullOrEmpty(turn.Assistant.Id)
                    || turn.Assistant.IdSource is null
                    || turn.Assistant.IdSource is not ("dom" or "dom-order")
                    || string.IsNullOrEmpty(turn.Assistant.Text))
                {
                    throw Error(CaptureValidationErrorCode.InvalidPayload, "Capture contains an invalid assistant message.");
                }

                if (!messageIds.Add(turn.Assistant.Id))
                {
                    throw Error(
                        CaptureValidationErrorCode.DuplicateMessageId,
                        $"Duplicate message ID: {turn.Assistant.Id}");
                }
            }

            if (turn.Complete && turn.Assistant is null)
            {
                throw Error(
                    CaptureValidationErrorCode.InconsistentTurn,
                    $"Turn {turn.Id} is marked complete without an assistant message.");
            }
        }
    }

    private static void ValidateSelection(CaptureSelection selection, IReadOnlyList<CapturedTurn> turns)
    {
        if (selection?.SelectedTurnIds is null
            || selection.SelectedTurnIds.Count != turns.Count
            || selection.SelectedTurnIds.Distinct(StringComparer.Ordinal).Count() != selection.SelectedTurnIds.Count
            || !selection.SelectedTurnIds.SequenceEqual(turns.Select(turn => turn.Id), StringComparer.Ordinal))
        {
            throw Error(
                CaptureValidationErrorCode.InvalidSelection,
                "Selected turn IDs must exactly match the payload turns.");
        }
    }

    private static byte[] SerializeCanonicalTurns(IReadOnlyList<CapturedTurn> turns)
    {
        using var stream = new MemoryStream();
        using (var writer = new Utf8JsonWriter(
                   stream,
                   new JsonWriterOptions
                   {
                       Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
                       Indented = false,
                   }))
        {
            writer.WriteStartArray();
            foreach (var turn in turns)
            {
                writer.WriteStartObject();
                if (turn.Assistant is not null)
                {
                    writer.WritePropertyName("assistant");
                    WriteMessage(writer, turn.Assistant);
                }

                writer.WriteBoolean("complete", turn.Complete);
                writer.WriteString("id", turn.Id);
                writer.WriteNumber("index", turn.Index);
                writer.WritePropertyName("user");
                WriteMessage(writer, turn.User);
                writer.WriteEndObject();
            }

            writer.WriteEndArray();
        }

        return stream.ToArray();
    }

    private static void WriteMessage(Utf8JsonWriter writer, CapturedMessage message)
    {
        writer.WriteStartObject();
        writer.WriteString("id", message.Id);
        writer.WriteString("idSource", message.IdSource);
        writer.WriteString("text", message.Text);
        writer.WriteEndObject();
    }

    private static Guid CreateStableGuid(string value)
    {
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(value));
        var hex = Convert.ToHexString(hash.AsSpan(0, 16));
        var valueWithSeparators = $"{hex[..8]}-{hex[8..12]}-{hex[12..16]}-{hex[16..20]}-{hex[20..32]}";
        return Guid.ParseExact(valueWithSeparators, "D");
    }

    private static CaptureValidationException Error(CaptureValidationErrorCode code, string message)
    {
        return new CaptureValidationException(code, message);
    }
}
